### PR TITLE
chore: script to migrate old backup files to new format

### PR DESCRIPTION
**Pull request by mihir20**
# Description
To run replay for duration when archival was not enabled, we have to run replay in older way which is very time consuming, With the help of this script we can convert old gateway backup files to new format archival files and run the replay through rudder-sources replay mechanism.

## Linear Ticket

https://linear.app/rudderstack/issue/PIPE-730/script-to-migrate-old-backup-files-to-new-archive-format

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: cmd/backupfilemigrator/file_migrator.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	var err error
+	dumpDownloadPathDirName := "/rudder-s3-dumps/"
+	tmpdirPath := strings.TrimSuffix(kitConfig.GetString("RUDDER_TMPDIR", "/tmp"), "/")
+	if tmpdirPath == "" {
+		tmpdirPath, err = os.UserHomeDir()
+		if err != nil {
+			return "", errors.Wrap(err, "failed to get user home directory")
+		}
+	}
+	path := fmt.Sprintf(
```
Comment (achettyiitr) [inline cmd/backupfilemigrator/file_migrator.go:177]: Variable 'path' collides with imported package name 
```suggestion
	tempPath := fmt.Sprintf(
```

File: cmd/backupfilemigrator/file_migrator.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+		`%v%v%v`, tmpdirPath, dumpDownloadPathDirName, filePathTokens[len(filePathTokens)-1])
+
+	err = os.MkdirAll(filepath.Dir(path), os.ModePerm)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to make local directory")
+	}
```
Comment (achettyiitr) [inline cmd/backupfilemigrator/file_migrator.go:183]:
```suggestion
	tmpDirPath, err := misc.CreateTMPDIR()
	if err != nil {
		return "", fmt.Errorf("creating tmp directory: %w", err)
	}

	dir, err := os.MkdirTemp(tmpDirPath, "*")
	if err != nil {
		return "", fmt.Errorf("making tmp directory: %w", err)
	}

	tempPath := path.Join(dir, filePathTokens[len(filePathTokens)-1])
```

File: cmd/backupfilemigrator/main.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+type newFileFormat struct {
+	UserID       string          `json:"userId"`
+	EventPayload json.RawMessage `json:"payload"`
+	CreatedAt    time.Time       `json:"createdAt"`
+	MessageID    string          `json:"messageId"`
+}
+
+// step 1: fetch files to migrate
+func main() {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply cmd/backupfilemigrator/main.go:31]: Are we planning to run this through Kubernetes jobs?

**Comment by achettyiitr** [comment]
**@mihir20 
[This](https://github.com/rudderlabs/rudder-server/blob/8ce20b440496f9aab4b28fdcfb6eb795aabb6d0d/cmd/backupfilemigrator/testdata/gw_jobs_52306.1311260446.1311289420.1703893976586.1703894226969.2HajBvyyfPW3n0QNSaYH7eEXt1z.gz) seems like real data from customer **cars**? Let not use actual customers data. We can get the dummy data from our dev environment as well or we can generate it as well?**

File: cmd/backupfilemigrator/file_migrator.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	return listOfFiles
+}
+
+// convert old file line entry to new file format
+func (m *fileMigrator) convertToNewFormat(lineBytes []byte, createdAt time.Time) ([]*newFileFormat, error) {
+	var gatewayBatchReq []types.SingularEventT
+	err := jsonfast.Unmarshal([]byte(gjson.GetBytes(lineBytes, "event_payload.batch").String()), &gatewayBatchReq)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to unmarshall gateway events")
```
[the remark this answers is not in the document]
Comment (mihir20) [reply cmd/backupfilemigrator/file_migrator.go:96]: Wrap felt clearer to me but didn't knew its not maintained any more.  will update 👍

File: cmd/backupfilemigrator/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mihir20) [reply cmd/backupfilemigrator/main.go:31]: yes, but we would need help from infra as resource consumption will be high to run this job

File: cmd/backupfilemigrator/main.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		reader, err := gzip.NewReader(rawf)
+		if err != nil {
+			log.Errorn("failed to get new gzip reader", obskit.Error(err))
+			continue // failed to read gzip file
+		}
+
+		sc := bufio.NewScanner(reader)
+		// default scanner buffer maxCapacity is 64K
+		// set it to higher value to avoid read stop on read size error
+		maxCapacity := 10240 * 1024 // 10MB
```
[the remark this answers is not in the document]
Comment (mihir20) [reply cmd/backupfilemigrator/main.go:123]: we are already keeping it a high value so I guess we should be good here, only down side here is we might be taking 10MBs of ram here which should be okay IMO.

**Comment by mihir20** [comment]
> @mihir20 [This](https://github.com/rudderlabs/rudder-server/blob/8ce20b440496f9aab4b28fdcfb6eb795aabb6d0d/cmd/backupfilemigrator/testdata/gw_jobs_52306.1311260446.1311289420.1703893976586.1703894226969.2HajBvyyfPW3n0QNSaYH7eEXt1z.gz) seems like real data from customer **cars**? Just wanted to make sure if it is, let's not push actual customers data in Github. We can get the dummy data from our dev environment or we can generate it as well?

This is dummy data only, not actual customer data. using ids or real looking data might have given this sense, so I have updated few fields with dummy data.

File: cmd/backupfilemigrator/file_migrator.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+	if err != nil {
+		return "", fmt.Errorf("failed to make local file: %w", err)
+	}
+
+	err = m.fileManager.Download(ctx, file, filePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to download file: %w", err)
+	}
+	m.logger.Debugf("file downloaded at %s", tempPath)
+	defer func() { _ = file.Close() }()
```
[the remark this answers is not in the document]
Comment (mihir20) [reply cmd/backupfilemigrator/file_migrator.go:203]: even if there is some error the best we can do is log the error, adding a log statement

**Review by achettyiitr — APPROVED** [review]
@mihir20 Great job with the PR 🚀

**Review by lvrach — APPROVED** [review]

**Pull request merged**